akka-cluster-tools/src/main/scala/akka/cluster/client/ClusterClient.scala

```diff
@@ -542,6 +568,28 @@ object ClusterReceptionist {
     Props(new ClusterReceptionist(pubSubMediator, settings)).withDeploy(Deploy.local)
 
   /**
+   * Declares a super type for all events emitted by the `ClusterReceptionist`.
+   * Events are published to the Akka event stream (system.eventStream).
+   */
+  sealed abstract class ClusterReceptionistEvent
+
+  /**
+   * Emitted to the Akka event stream when a heartbeat has been
+   * received from a ClusterClient. Consumers of the event should expect
+   * to receive it periodically and at reasonably regular intervals.
+   *
+   * It is up to the receiver of this event to discern whether there are new
+   * cluster clients or whether the cluster clients have been removed since any
+   * prior `HeartbeatReceived` event.
+   *
+   * @param clusterClient The cluster client that sent the heartbeat
+   * @param clusterReceptionist The receptionist that emitted the event
+   */
+  final case class HeartbeatReceived(
+    clusterClient: ActorPath,
+    clusterReceptionist: ActorRef) extends ClusterReceptionistEvent
```

Comment:
We originally had the worker register its availability via a message. However in order to track its continued availability then the worker must send such a message periodically - a heartbeat of sorts. Hence the solution here leveraging the existing heartbeat mechanism. 

An alternate solution would be for the server to deathwatch the client. Is that possible?

> On 6 May 2016, at 08:15, Patrik Nordwall <notifications@github.com> wrote:
> 
> In akka-cluster-tools/src/main/scala/akka/cluster/client/ClusterClient.scala:
> 
> > +
> > +  /**
> > +   * Emitted to the Akka event stream when a heartbeat has been
> > +   * received from a ClusterClient. Consumers of the event should expect
> > +   * to receive it periodically and at reasonably regular intervals.
> > +   *
> > +   * It is up to the receiver of this event to discern whether there are new
> > +   * cluster clients or whether the cluster clients have been removed since any
> > +   * prior `HeartbeatReceived` event.
> > +   *
> > +   * @param clusterClient The cluster client that sent the heartbeat
> > +   * @param clusterReceptionist The receptionist that emitted the event
> > +   */
> > +  final case class HeartbeatReceived(
> > +    clusterClient: ActorPath,
> > +    clusterReceptionist: ActorRef) extends ClusterReceptionistEvent
> I don't think that should be solved via the heartbeats. Shouldn't your worker publish a message via the cluster client and then you can act on that in the destination actor.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>